les/catalyst/api.go

```diff
@@ -66,124 +78,222 @@ func NewConsensusAPI(les *les.LightEthereum) *ConsensusAPI {
+		var (
+			td  = api.les.BlockChain().GetTd(update.HeadBlockHash, header.Number.Uint64())
+			ptd = api.les.BlockChain().GetTd(header.ParentHash, header.Number.Uint64()-1)
+			ttd = api.les.BlockChain().Config().TerminalTotalDifficulty
+		)
+		if td == nil || (header.Number.Uint64() > 0 && ptd == nil) {
+			log.Error("TDs unavailable for TTD check", "number", header.Number.Uint64(), "hash", update.HeadBlockHash, "td", td, "parent", header.ParentHash, "ptd", ptd)
+			return beacon.STATUS_INVALID, errors.New("TDs unavailable for TDD check")
+		}
+		if td.Cmp(ttd) < 0 {
+			log.Error("Refusing beacon update to pre-merge", "number", header.Number.Uint64(), "hash", update.HeadBlockHash, "diff", header.Difficulty, "age", common.PrettyAge(time.Unix(int64(header.Time), 0)))
+			return beacon.ForkChoiceResponse{PayloadStatus: beacon.INVALID_TERMINAL_BLOCK, PayloadID: nil}, nil
+		}
+		if header.Number.Uint64() > 0 && ptd.Cmp(ttd) >= 0 {
+			log.Error("Parent block is already post-ttd", "number", header.Number.Uint64(), "hash", update.HeadBlockHash, "diff", header.Difficulty, "age", common.PrettyAge(time.Unix(int64(header.Time), 0)))
+			return beacon.ForkChoiceResponse{PayloadStatus: beacon.INVALID_TERMINAL_BLOCK, PayloadID: nil}, nil
+		}
+	}
+	valid := func(id *beacon.PayloadID) beacon.ForkChoiceResponse {
+		return beacon.ForkChoiceResponse{
+			PayloadStatus: beacon.PayloadStatusV1{Status: beacon.VALID, LatestValidHash: &update.HeadBlockHash},
+			PayloadID:     id,
+		}
+	}
+	if rawdb.ReadCanonicalHash(api.les.ApiBackend.ChainDb(), header.Number.Uint64()) != update.HeadBlockHash {
+		// Block is not canonical, set head.
+		if err := api.les.BlockChain().SetCanonical(header); err != nil {
+			return beacon.ForkChoiceResponse{PayloadStatus: beacon.PayloadStatusV1{Status: beacon.INVALID, LatestValidHash: nil}}, err
+		}
+	} else if api.les.BlockChain().CurrentHeader().Hash() == update.HeadBlockHash {
+		// If the specified head matches with our local head, do nothing and keep
+		// generating the payload. It's a special corner case that a few slots are
+		// missing and we are requested to generate the payload in slot.
+	} else {
+		// If the head block is already in our canonical chain, the beacon client is
+		// probably resyncing. Ignore the update.
+		log.Info("Ignoring beacon update to old head", "number", header.Number.Uint64(), "hash", update.HeadBlockHash, "age", common.PrettyAge(time.Unix(int64(header.Time), 0)), "have", api.les.BlockChain().CurrentHeader().Number.Uint64())
+		return valid(nil), nil
+	}
+
+	// If the beacon client also advertised a finalized block, mark the local
+	// chain final and completely in PoS mode.
+	if update.FinalizedBlockHash != (common.Hash{}) {
+		if merger := api.les.Merger(); !merger.PoSFinalized() {
+			merger.FinalizePoS()
 		}
+		// If the finalized block is not in our canonical tree, somethings wrong
+		finalBlock, err := api.les.BlockChain().GetBlockByHash(ctx, update.FinalizedBlockHash)
```

Comment:
Please don't use `GetBlockByHash` which will retrieve the block from LES server if block is non-existent.